torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -127,11 +125,26 @@ def powerSGD_hook(
     input_tensor[total_length:padded_total_length].fill_(0)
 
     # Incorporate the error from the previous state into the gradients.
+    bucket_index = bucket.get_index()
     if state.use_error_feedback:
-        if input_tensor in state.error_dict:
-            input_tensor.add_(state.error_dict[input_tensor])
+        # The buckets can be rebuilt during training.
+        # In this case, the error tensor shape will not be aligned with the input tensor,
+        # and the error will be re-initialized as zeros.
+        if (
+            bucket_index in state.error_dict
+            and state.error_dict[bucket_index].shape[0] == padded_total_length
+        ):
+            input_tensor.add_(state.error_dict[bucket_index])
         else:
-            state.error_dict[input_tensor] = torch.zeros(padded_total_length, device=device)
+            logging.info(
```

Comment:
Even though we rebuild buckets only once, is it detrimental to power SGD if we have to restart from zeros after rebuilding buckets? If it might be, here are some ideas:
1) Keep track of the average error and substitute that in
2) Maintain a mapping of old index to new index (since the indices can change with rebuilt buckets), and transfer over the error appropriately when it is determined in this hook that buckets were rebuilt. Although, this might make the hook rely too much on the details of an internal DDP implementation (which feels like overall, a general downside of directly using GradBuckets in the hook?)

Reply:
It should be fine as long as such mismatch does not happen consistently. Training is an iterative process, so even the error values are wrong occasionally, the later training should be able to fix that.